Motion tweaking

starwed opened the pull request:
Several small issues with Motion/Control components
- Fix bugs with NewDirection (and angular version)
  - Fixes a rounding issue when calculating the sign of negative fractions
  - Triggers NewDirection in the frame where the direction changes, instead of the next frame
  - Renames NewRotation to NewAngularDirection and fixes the same issues
  - Adds/fixes related tests
- Sync motion with clock rather than number of frames
  - Instead of adjusting the speed based on the frame rate, try to saty in sync with the game clock by using frame.dt.  Note that speeds are in pixels per second.
  - This is a **breaking change**, because it changes how we interpret speeds.  (Pixels per second instead of pixels per frame.)
  - Since it's a breaking change, I needed to fix several tests.  I found the existing tests very hard to follow, so I added a lot of clarification, and removed some that seemed redundant.
- Change Jumpway -> Jumper
-  Reword motion docs

starwed committed:
Reword motion docs

Add some in-line documentaiton, and tweak the wording of the API docs for the Motion component

starwed committed:
Fix bugs with NewDirection (and angular version)

- Fixes a rounding issue when calculating the sign of negative fractions
- Triggers NewDirection in the frame where the direction changes, instead of the next frame
- Renames NewRotation to NewAngularDirection and fixes the same issues
- Adds/fixes related tests

starwed committed:
Sync motion with clock rather than number of frames

Instead of adjusting the speed based on the frame rate, try to saty in sync with the game clock by using frame.dt.  Note that speeds are in pixels per second.

This is a breaking change, because it changes how we interpret speeds.  (Pixels per second instead of pixels per frame.)

starwed committed:
Change Jumpway -> Jumper

"Jumpway" is probably a bit confusing, even though it kind of makes sense in the context of Twoway and Fourway.

starwed commented:
As mentioned, this does have a breaking change: it changes speeds from pixels per frame to pixels per second.

I think this is a more natural definition in most cases -- I often had to choose very small speeds or accelerations with the old units.  Any arguments to  keep it as is, though?

starwed commented:
A quick note about tests: lots of tests are certainly way better than none!  But I wanted to give a few recommendations.  (I'm not trying to pick on the tests for Motion, they just happened to be what I happened to interface with on this patch!  And the thorough testing definitely let me feel good about making the changes!  :)  )
- If tests don't have a description string, it becomes a bit difficult to discover which specific test is failing.
- Ideally, each `test()` block should test one set of things in reasonable isolation, unless it's _supposed_ to be an integration test.  (Those are good too!)
- There's no need to test every aspect of a behavior.  (For instance, a lot of the tests for multiway checked that the velocity was set correctly _and_ that the entities moved the correct distance -- the latter is actually testing "Motion", not "Multiway".)

mucaho commented on tests/2d.js:
```diff
@@ -562,106 +624,105 @@
-    equal(newRevolutionEvents, 3);
-    equal(movedEvents, 6);
-    equal(rotatedEvents, 2);
-    equal(motionEvents, 17);
```
this checks shouldn't be removed, as we can not guarantee that all event callbacks fired as we wanted

mucaho commented:
I have looked over the PR, good job!
Specifying speed in pixels per second is the most groundbreaking change I have witnessed so far. It's the right thing to do, but this will probably break all existing games! That's why I initially proposed we stick to `dt * 0,05`.
Nice job identifying the problem with the `Math.sign` solution from [SO](http://stackoverflow.com/a/23262976/3041008).

The only thing I would like to nitpick is the order of the `NewDirection` event. The motivation behind it was that `NewDirection` should be fired just before the velocity/acceleration was applied. If user has custom `EnterFrame` callback that changes velocity/acceleration, he can not be sure whether it is applied before/after `Motion` changes the velocity/acceleration. By binding to the `NewDirection` event he can make sure what combined velocity/accel will be applied and still change it to suit his needs (e.g. he can prevent moving left).
Initially I had the `NewDirection` event fire after `Motion` changed it (as it's natural), but maybe there is a corner case later where the paragraph above will be relevant. We can leave it as proposed in this PR for now.

starwed commented:
> Initially I had the NewDirection event fire after Motion changed it (as it's natural), but maybe there is a corner case later where the paragraph above will be relevant. We can leave it as proposed in this PR for now.

Another possibility would just be to trigger this when the velocity is actually changed.  (So in the setter for the velocity properties.)  That way it wouldn't matter _how_ the velocity was changed.

starwed commented:
> Specifying speed in pixels per second is the most groundbreaking change I have witnessed so far. It's the right thing to do, but this will probably break all existing games! That's why I initially proposed we stick to dt \* 0,05.

I think there are enough breaking changes since the last version that pretty much any game will require a bit of work to update, so I'm not too worried about this.  I find it makes it easier to reason about how speeds and accelerations, so it should be worth it.

mucaho commented:
>  I find it makes it easier to reason about how speeds and accelerations, so it should be worth it.

It also plays nicely with tweening.

Ok then, I suggest you readd [some test checks](https://github.com/craftyjs/Crafty/pull/923#discussion_r32488655), as well as update the default values for movement and their respective documentation (`_speed` in `Multiway` and `_jumpSpeed` in `Jumper`).
After that we can merge this :)

mucaho commented:
> Another possibility would just be to trigger this when the velocity is actually changed. (So in the setter for the velocity properties.) That way it wouldn't matter how the velocity was changed.

Yes, that would be great! Note that it will introduce some overhead probably.

starwed commented:
Ok, I re-added those tests.  (A better fix would be to break up that long litany of tests into several isolated ones, and then check that each event was called individually, but I think this is fine for now.)

I also adjusted the default speeds/accelerations to match the new time-scale.

mucaho commented on src/controls/controls.js:
```diff
@@ -450,13 +451,13 @@ Crafty.c("Jumpway", {
      * ~~~
-     * this.jumpway(6, ['UP_ARROW', 'W']);
-     * this.jumpway(['UP_ARROW', 'W']);
+     * this.jumper(6, ['UP_ARROW', 'W']);
```
update speed here, e.g. `300`

mucaho commented on src/controls/controls.js:
```diff
@@ -470,14 +471,14 @@ Crafty.c("Jumpway", {
+     * @comp Jumper
      * @sign public this .jumpSpeed(Number jumpSpeed)
      * @param jumpSpeed - new vertical jump speed
      *
```
update speed here in example below

mucaho commented on src/controls/controls.js:
```diff
@@ -234,7 +234,8 @@ Crafty.c("Multiway", {
+     *
      * Can be called while a key is pressed to change speed on the fly.
      *
      * @example
```
update speed here in example below

mucaho commented on src/controls/controls.js:
```diff
@@ -185,7 +185,7 @@ Crafty.c("Multiway", {
+     * @param speed - A speed in pixels per second
      * @param keyBindings - What keys should make the entity go in which direction. Direction is specified in degrees
      *
      * Constructor to initialize the speed and keyBindings. Component will listen to key events and move the entity appropriately.
```
update speed here in example below

mucaho commented on src/spatial/2d.js:
```diff
@@ -1182,9 +1184,9 @@ Crafty.c("Gravity", {
+     * @param g - gravitational constant in pixels per second squared
      *
-     * Set the gravitational constant to g. The default is 0.2 . The greater g, the faster the object falls.
+     * Set the gravitational constant to g for this entity. The default is 50. The greater g, the stronger the downwards acceleration.
```
default is `500`

mucaho commented:
Great, I noticed some minor things that should be updated in documentation.

> Another possibility would just be to trigger this when the velocity is actually changed. (So in the setter for the velocity properties.) That way it wouldn't matter how the velocity was changed.

Should this be in current or seperate PR?

starwed commented:
Good catch with the examples.

> Should this be in current or seperate PR?

If we do that, it'll be a separate PR.  Since there's an event any time the velocity changes, I'm not sure how important it is to have a separate one that occurs for a particular type of change.

starwed commented:
Ok, all the examples should be updated now.

letmaik commented:
I just noticed this since I use the nightlies:
http://neothemachine.github.io/hungrymonkey/

I think to make the transition easier it would good to provide some rough formula to get from pixels per frame to pixels per second, assuming some common framerates like 60fps.

mucaho commented:
To go from old `[px / frame]` to new `[px / s]` you multiply it by the target frame rate, which defaults to `50`, e.g.:
- old `jumpSpeed = 6 [px / frame]` -> new `jumpSpeed = 300 [px / s]`
- old `gravityConst = 0.2 [px / frame²]` -> new `gravityConst = 0.2 * 50² = 500 [px / s²]`

Good idea, we should probably add this to the changelog

letmaik commented:
Yep, that worked. Thanks
